From 182f613d13a6df552b1fb7da73a7fab95d024e89 Mon Sep 17 00:00:00 2001
From: Harrison Yang <harrison.yang@siliconmotion.com>
Date: Fri, 1 Aug 2025 14:06:26 +0800
Subject: [PATCH] Purpose: Make Obmc-iKVM can work with Raspberry Pi 4 HID.

Reoslution: The gadget address of Raspberry Pi 4 is fe980000.usb.
Upstream-Status: Submitted

Reviewed-by: None
Signed-off-by: Harrison Yang <harrison.yang@siliconmotion.com>
---
 create_usbhid.sh  | 19 +++++--------------
 obmc-ikvm.service |  2 +-
 2 files changed, 6 insertions(+), 15 deletions(-)

diff --git a/create_usbhid.sh b/create_usbhid.sh
index 58e4034..28db072 100644
--- a/create_usbhid.sh
+++ b/create_usbhid.sh
@@ -1,7 +1,7 @@
 #!/bin/sh
 
 hid_conf_directory="/sys/kernel/config/usb_gadget/obmc_hid"
-dev_name="1e6a0000.usb-vhub"
+dev_name="fe980000.usb"
 
 create_hid() {
     # create gadget
@@ -125,28 +125,19 @@ create_hid() {
 }
 
 connect_hid() {
-    if ! grep -q "${dev_name}:p" UDC; then
-        i=0
-        num_ports=5
-        base_usb_dir="/sys/bus/platform/devices/${dev_name}/${dev_name}:p"
-        while [ "${i}" -lt "${num_ports}" ]; do
-            port=$(("${i}" + 1))
-            i="${port}"
-            if [ ! -e "${base_usb_dir}${port}/gadget/suspended" ]; then
-                break
-            fi
-        done
-        echo "${dev_name}:p${port}" > UDC
+    if ! grep -q "${dev_name}" UDC; then
+        echo "${dev_name}" > UDC
     fi
 }
 
 disconnect_hid() {
-    if grep -q "${dev_name}:p" UDC; then
+    if grep -q "${dev_name}" UDC; then
         echo "" > UDC
     fi
 }
 
 if [ ! -e "${hid_conf_directory}" ]; then
+    modprobe -a tc358743 bcm2835_unicam i2c_mux_pinctrl libcomposite dwc2
     create_hid
 else
     cd "${hid_conf_directory}" || exit 1
diff --git a/obmc-ikvm.service b/obmc-ikvm.service
index 60234b2..687341b 100644
--- a/obmc-ikvm.service
+++ b/obmc-ikvm.service
@@ -5,7 +5,7 @@ ConditionPathIsMountPoint=/sys/kernel/config
 [Service]
 Restart=always
 ExecStartPre=/usr/bin/create_usbhid.sh disconnect
-ExecStart=/usr/bin/obmc-ikvm -v /dev/video0 -k /dev/hidg0 -p /dev/hidg1
+ExecStart=/usr/bin/obmc-ikvm -v /dev/video0 -k /dev/hidg0 -p /dev/hidg1 -u fe980000.usb
 
 [Install]
 WantedBy=multi-user.target
